### DISCUSSION TITLE
Project prototype and goals

### DISCUSSION THREAD

**Discussion by nyluntu** [sovelluskontti/project-semantic-search#52] [General]
Hi.

I will write down some ideas on what the final prototype could look like. Let's discuss this today.

# Frontend

The prototype should consist of a simple frontend page accessible through a browser. Users can test different search queries and methods to view the results. Initially, the results should contain only movie titles.
 
![search page](https://github.com/user-attachments/assets/4cd11b39-1049-4e3b-bda0-8c5f1ede5ed7)

# Backend

The backend should include two endpoints to handle search queries and return results to the frontend. The search is performed by Manticore Search.

The backend does not need to handle any other functionalities. Data for Manticore Search can be prepared for the database in any way needed.

## Keyword Search

The backend should send the provided keywords to the search engine, using only basic keyword search functionality. This means that the given word or set of words should be present in the movie title.

## Semantic Search

The backend should send the search query to the search engine. The input query should first be converted to embeddings before performing the search.

# Technology stack

I suggest using the following stack:

- Frontend: React
- Backend: Flask
- Database: Manticore Search

These tools should cover everything needed for the project. Let's aim to keep things simple and somewhat familiar where possible. The main focus is on the search functionalities rather than a polished frontend. My goal is to see all components working together.

We'll use Docker to set up the development environment. The Docker setup should include the following services:

- A container for the React frontend
- A container for the Flask backend
- A container for the Manticore Search database
- Possibly a container for an embedding generator, or we could use an external API. I’m uncertain if a local embedding solution is ideal, as such tools often require significant resources. An external API might be preferable to keep the containers lightweight, which would also improve the local development experience.

For this phase of the project, it would be best to start by setting up the development environment. After that, we can implement the keyword search functionality. This approach will likely give us more clarity on how to handle the semantic search portion.

**Comment by nyluntu** [comment]
@sarafarahabadi Let's discuss this in our upcoming meeting. We need to determine whether we’re ready to start building the entire prototype or if additional research is needed first. If we agree to move forward, we’ll also need to redefine the backlog.

**Reply by sarafarahabadi** [reply]
Hello Turo, 

Thanks for the detailed info. I was searching for which technologies to use and I agree with you. I will update the project backlog based on the project's current course. It would also be simpler for me to start the frontend for keyword search and then move on to the semantic part.
